File: fittracker-frontend/src/api/templates.ts
import { apiClient } from './client';

export interface TemplateExercise {
  id: number;
  exercise_library_id: number;
  exercise_name: string | null;
  exercise_order: number;
  target_sets: number | null;
  target_reps: number | null;
  target_weight_kg: number | null;
  rest_seconds: number;
  notes: string | null;
}

export interface Template {
  id: number;
  user_id: number;
  name: string;
  description: string | null;
  is_public: boolean;
  estimated_duration_min: number | null;
  times_used: number;
  created_at: string;
  exercise_count: number;
}

export interface TemplateDetail extends Template {
  exercises: TemplateExercise[];
  creator_username: string | null;
}

export interface TemplateCreate {
  name: string;
  description?: string;
  is_public?: boolean;
  estimated_duration_min?: number;
  exercises?: Array<{
    exercise_library_id: number;
    exercise_order?: number;
    target_sets?: number;
    target_reps?: number;
    target_weight_kg?: number;
    rest_seconds?: number;
    notes?: string;
  }>;
}

export const templatesApi = {
  list: (params: { mine?: boolean; public?: boolean } = {}): Promise<Template[]> => {
    const q = new URLSearchParams();
    if (params.mine !== undefined) q.set('mine', String(params.mine));
    if (params.public !== undefined) q.set('public', String(params.public));
    return apiClient.get<Template[]>(`/api/templates?${q}`).then(r => r.data);
  },
  get: (id: number): Promise<TemplateDetail> =>
    apiClient.get<TemplateDetail>(`/api/templates/${id}`).then(r => r.data),
  create: (data: TemplateCreate): Promise<TemplateDetail> =>
    apiClient.post<TemplateDetail>('/api/templates', data).then(r => r.data),
  update: (id: number, data: Partial<TemplateCreate>): Promise<TemplateDetail> =>
    apiClient.put<TemplateDetail>(`/api/templates/${id}`, data).then(r => r.data),
  delete: (id: number): Promise<void> =>
    apiClient.delete(`/api/templates/${id}`).then(() => undefined),
  start: (id: number): Promise<{ id: number; name: string }> =>
    apiClient.post(`/api/templates/${id}/start`).then(r => r.data),
  saveAsTemplate: (workoutId: number, data: { name: string; description?: string; is_public?: boolean }): Promise<Template> =>
    apiClient.post<Template>(`/api/workouts/${workoutId}/save-as-template`, data).then(r => r.data),
};

File: fittracker-backend/app/schemas/templates.py
from __future__ import annotations
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, field_validator


class TemplateExerciseCreate(BaseModel):
    exercise_library_id: int
    exercise_order: int = 1
    target_sets: Optional[int] = None
    target_reps: Optional[int] = None
    target_weight_kg: Optional[float] = None
    rest_seconds: int = 60
    notes: Optional[str] = None


class TemplateExerciseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    exercise_library_id: int
    exercise_order: int
    target_sets: Optional[int] = None
    target_reps: Optional[int] = None
    target_weight_kg: Optional[float] = None
    rest_seconds: int
    notes: Optional[str] = None
    exercise_name: Optional[str] = None  # populated manually


class TemplateCreate(BaseModel):
    name: str
    description: Optional[str] = None
    is_public: bool = False
    estimated_duration_min: Optional[int] = None
    exercises: list[TemplateExerciseCreate] = []

    @field_validator('name')
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError('Name cannot be empty')
        return v.strip()


class TemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_public: Optional[bool] = None
    estimated_duration_min: Optional[int] = None


class TemplateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: int
    name: str
    description: Optional[str] = None
    is_public: bool
    estimated_duration_min: Optional[int] = None
    times_used: int
    created_at: datetime
    exercise_count: int = 0  # populated manually


class TemplateDetailOut(TemplateOut):
    exercises: list[TemplateExerciseOut] = []
    creator_username: Optional[str] = None  # populated manually


class TemplateSaveFromWorkout(BaseModel):
    name: str
    description: Optional[str] = None
    is_public: bool = False
    estimated_duration_min: Optional[int] = None
